### PR TITLE
Use fsGroup and handles NFS shares

### LINKED ISSUE #357: Permission denied when using persistent volume AzureDisk
##### ISSUE TYPE
 - Bug Report

##### SUMMARY
I am getting permission denied when trying to clone git into /var/lib/awx/projects when using persistent storage over AzureDisk

##### ENVIRONMENT
* AWX version: 19.1.0
* Operator version: 0.10.0
* Kubernetes version: 1.18.14
* AWX install method: Azure Kubernetes services

##### STEPS TO REPRODUCE
1) Create PV and PVC with defined storage class
```
---
apiVersion: storage.k8s.io/v1
kind: StorageClass
allowVolumeExpansion: true
metadata:
  name: azuredisk
  labels:
    addonmanager.kubernetes.io/mode: EnsureExists
    kubernetes.io/cluster-service: "true"
parameters:
  cachingmode: ReadOnly
  kind: Managed
  storageaccounttype: Premium_LRS
provisioner: kubernetes.io/azure-disk
reclaimPolicy: Retain
volumeBindingMode: Immediate

---

apiVersion: v1
kind: PersistentVolume
metadata:
  name: pv-awx
  namespace: awx
spec:
  capacity:
    storage: 64Gi
  accessModes:
    - ReadWriteOnce
  storageClassName: azuredisk
  claimRef:
    namespace: awx
    name: pvc-awx
  azureDisk:
    kind: Managed
    diskName: disk-pv-awx
    diskURI: /subscriptions/XXXXXXXX/resourceGroups/RGNAME/providers/Microsoft.Compute/disks/disk-pv-awx

---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: pvc-awx
  namespace: awx
spec:
  accessModes:
    - ReadWriteOnce
  storageClassName: default-retain
  resources:
    requests:
      storage: 64Gi
```

2) Create AWX using this deployment
```
---

apiVersion: awx.ansible.com/v1beta1
kind: AWX
metadata:
  name: awx
  namespace: awx
spec:
  ingress_type: Ingress
  hostname: awx.hostname
  ingress_annotations: |
    cert-manager.io/cluster-issuer: letsencrypt-prod
    kubernetes.io/ingress.class: nginx
  ingress_tls_secret:
    awx-le-secret
  projects_existing_claim: pvc-awx
  projects_persistence: true
```

##### EXPECTED RESULTS
Permissions of mounted volume are set properly and clone of git works well

##### ACTUAL RESULTS
I am getting permission denied as projects folder is mounted 755 and owner is root so runtime user is not able to write into that folder

```
Traceback (most recent call last):
  File "/var/lib/awx/venv/awx/lib64/python3.8/site-packages/awx/main/tasks.py", line 1289, in run
    self.pre_run_hook(self.instance, private_data_dir)
  File "/var/lib/awx/venv/awx/lib64/python3.8/site-packages/awx/main/tasks.py", line 2144, in pre_run_hook
    os.makedirs(project_path)  # used as container mount
  File "/usr/lib64/python3.8/os.py", line 223, in makedirs
    mkdir(name, mode)
PermissionError: [Errno 13] Permission denied: '/var/lib/awx/projects/_8__ansible'
```
##### ADDITIONAL INFORMATION
Using azure files doesn't work also as AzureFiles doesn't support chmod.

##### AWX-OPERATOR LOGS
N/A

### LINKED ISSUE #395: Not able to write to NFS volume mount
Hello guys, I'm having an issue that seems to be easy to fix but somehow everything I try is not working. The NFS volumes are mounting properly but the awx user is not able to write to them only read. Even after running the pods in privileged mode I still can't write to the mount points. I tried using securityContext but AWX does not support it.
image
image

```
# storage.yml needs to be ran first
# database_import_secret.yml needs to be ran first as well
---
apiVersion: awx.ansible.com/v1beta1
kind: AWX
metadata:
  name: awx
  namespace: awx
spec:
  hostname: test
  old_postgres_configuration_secret: awx-old-postgres-configuration
  admin_email: test
  task_privileged: true
  projects_persistence: true
  projects_storage_class: oci
  projects_storage_size: 50gb
  projects_storage_access_mode: ReadWriteOnce
  projects_existing_claim: oke-ocipvc-awx-projects
  extra_volumes: |
                      - name: oke-fsspvc-awx-scratch
                        persistentVolumeClaim:
                          claimName: oke-fsspvc-awx-scratch
  ee_extra_volume_mounts:  |
                      - name: oke-fsspvc-awx-scratch
                        persistentVolumeClaim:
                          claimName: oke-fsspvc-awx-scratch
                        mountPath: /scratch
```

```
# Create Persistent Volume (PV) - For /scatch

apiVersion: v1
kind: PersistentVolume
metadata:
  name: oke-fsspv-awx-scratch
spec:
  storageClassName: oci-fss
  capacity:
    storage: 100Gi
  accessModes:
    - ReadWriteMany
  mountOptions:
    - nosuid
  nfs:
    server: 10.50.132.5 #DNS names cant be used atm since pods are not able to resolve the dns
    path: "/inf_lower_awx_testing_scratch"
    readOnly: false

```

```
# Crate Persistent Volume Claim (PVC) - For /scratch

apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: oke-fsspvc-awx-scratch
  namespace: awx
spec:
  storageClassName: oci-fss
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      # Although storage is provided here it is not used for FSS file systems
      storage: 10Gi
  volumeName: oke-fsspv-awx-scratch

```

### PR DESCRIPTION
**Pull request by tchellomello**
Fixes: https://github.com/ansible/awx-operator/issues/395
Fixes: https://github.com/ansible/awx-operator/issues/357

This PR accomplishes a couple of things based on different environments. 

When using the `projects_persistence` with a `projects_existing_claim`, depending on the storage provide a `block` device or `filesystem` device will present to the AWX instance writable only by the `root` user.  This introduces a problem as the `awx` user is unable to write under `/var/lib/awx/projects` therefore not being able to sync projects. 

Using the `fsGroup` address the issue for `block` and `fs` mounts so the volume will be mounted allowing the `awx` user to write on it. However, `fsGroup` does not work for `NFS` shares as noted here: https://github.com/kubernetes/examples/issues/260

To work around this issue, we can use the `initContainer` approach to change the permissions and privileges for the mounted volume. 

Please refer to https://github.com/ansible/awx-operator/issues/395#issuecomment-864453580 and https://github.com/ansible/awx-operator/issues/357#issuecomment-863732211 for a full write-up while troubleshooting this issue. 

Another important addition is the ability to the user specify extra volumes as mentioned at https://github.com/ansible/awx-operator/issues/395#issuecomment-864659530

For example:

```yaml

spec:
  replicas: 1 
  ingress_type: none
  service_type: loadbalancer
  projects_existing_claim: awx-testing-nfs-claim
  projects_persistence: true
  extra_volumes: |
    - name: shared-volume
      persistentVolumeClaim:
        claimName: awx-testing-nfs-claim2
  init_container_extra_volume_mounts: |
    - name: shared-volume
      mountPath: /shared
  init_container_extra_commands: |
    # this will ensure the awx user can write on this volume
    chmod 775 /shared
    chgrp 1000 /shared
    date >> /var/lib/awx/projects/been_here
  task_extra_volume_mounts: |
    - name: shared-volume
      mountPath: /shared
```

**TL;DR**

**Before with nfs**
```sh
kubectl iexec awx /bin/bash                                                               15:08:49
Namespace: default | Pod: ✔ awx-nfs-596f6cf8b7-lb4xq
Container: ✔ awx-nfs-web
bash-4.4$ ls -lahd /var/lib/awx/projects
d---------. 1 root root 14 Jun 19 18:43 /var/lib/awx/projects
bash-4.4$ df -h /var/lib/awx/projects
Filesystem                        Size  Used Avail Use% Mounted on
192.168.111.46:/volume1/test-nfs-share  3.5T  1.3T  2.3T  36% /var/lib/awx/projects
bash-4.4$ whoami
awx
bash-4.4$ id awx
uid=1000(awx) gid=0(root) groups=0(root)
bash-4.4$ touch /var/lib/awx/projects/testing
touch: cannot touch '/var/lib/awx/projects/testing': Permission denied
```
**Before with block**
```sh
(py39) mdemello@storm ~/d/P/a/reproducer-357 (devel)> kubectl iexec awx-az /bin/bash                                       22:39:54
Namespace: default | Pod: ✔ awx-az-84dd57946f-855mn
Container: ✔ awx-az-web
bash-4.4$ df -h /var/lib/awx/projects/
Filesystem      Size  Used Avail Use% Mounted on
/dev/sdc        9.8G   37M  9.8G   1% /var/lib/awx/projects
bash-4.4$ ls -lah /var/lib/awx/projects/
total 24K
drwxr-xr-x 3 root root 4.0K Jun 18 02:26 .
drwxrwxr-x 1 root root 4.0K Jun 18 02:39 ..
drwx------ 2 root root  16K Jun 18 02:26 lost+found
bash-4.4$ id awx
uid=1000(awx) gid=0(root) groups=0(root)
bash-4.4$ whoami 
awx
bash-4.4$ touch /var/lib/awx/projects/test
touch: cannot touch '/var/lib/awx/projects/test': Permission denied
```

**After with nfs**
```sh
$ kubectl iexec awx /bin/bash                                                                   15:27:14
Namespace: default | Pod: ✔ awx-nfs-696dd649f5-zlqc4
Container: ✔ awx-nfs-web
bash-4.4$ df -h /var/lib/awx/projects/
Filesystem                        Size  Used Avail Use% Mounted on
192.168.111.46:/volume1/test-nfs-share  3.5T  1.3T  2.3T  36% /var/lib/awx/projects
bash-4.4$ whoami 
awx
bash-4.4$ id
uid=1000(awx) gid=0(root) groups=0(root),1000
bash-4.4$ touch /var/lib/awx/projects/testing-works-now
bash-4.4$ ls -la /var/lib/awx/projects/testing-works-now
-rw-rw-rw-. 1 awx root 0 Jun 19 19:23 /var/lib/awx/projects/testing-works-now
```
**After with block**

```sh
$ kubectl iexec awx-az /bin/bash                                                                            23:39:31
Namespace: default | Pod: ✔ awx-az-5fc5dd5767-lc48z
Container: ✔ awx-az-web
bash-4.4$ ls -lad /var/lib/awx/projects/
drwxrwsr-x 5 root 1000 4096 Jun 18 03:09 /var/lib/awx/projects/
bash-4.4$ df -h /var/lib/awx/projects/
Filesystem      Size  Used Avail Use% Mounted on
/dev/sdc        9.8G   37M  9.8G   1% /var/lib/awx/projects
bash-4.4$ touch /var/lib/awx/projects/test_works && echo $?
0
```

TODO:
* [ ] Tested @ AWS
* [ ] Tested @ OCP
* [x] Tested @ Azure :heavy_check_mark: 
* [x] Tested @ vanilla k8s with `hostPath` 
* [x] Tested @ vanilla k8s
* [x] Tested @ vanilla k8s with NFS provisioner

### CODE REVIEW

**Comment by felipe4334** [comment]
@tchellomello do you know if this changes have been merged with 0.12.0?

**Comment by tchellomello** [comment]
@felipe4334 no, it's not in the `0.12.0`. I'll finish testing this PR in some other environment and flip to get an official review. It should be included on the next version. :crossed_fingers:

**Comment by wohnout** [comment]
Thanks @tchellomello. This pull request helps a lot !

File: roles/installer/templates/deployment.yaml.j2 [hunk 1]
```diff
@@ -282,6 +300,10 @@ spec:
 {% if tolerations %}
       tolerations:
         {{ tolerations | indent(width=8) }}
+{% endif %}
+{% if projects_persistence|bool %}
+      securityContext:
+        fsGroup: 1000
```
Comment (shanemcd) [inline roles/installer/templates/deployment.yaml.j2:306]: I dont think this is correct. The UID in OpenShift is going to be unpredictable:

```
sh-4.4$ id -u
1000720000
```

Comment (shanemcd) [reply roles/installer/templates/deployment.yaml.j2:306]: The correct way to handle permissions in OCP is via the `root` group.

Some links that may be helpful: 

https://www.openshift.com/blog/a-guide-to-openshift-and-uids

See "Support Arbitrary User IDs" under https://docs.openshift.com/container-platform/3.11/creating_images/guidelines.html#openshift-specific-guidelines

Comment (shanemcd) [reply roles/installer/templates/deployment.yaml.j2:306]: > Notice the Container is using the UID from the Namespace. An important detail to notice is that the user in the Container always has GID=0, which is the root group.

**Review by shanemcd — APPROVED** [review]

File: roles/installer/templates/deployment.yaml.j2 (hunk 1 above)
Comment (tchellomello) [reply roles/installer/templates/deployment.yaml.j2:306]: Default SCC on openshift 4.7 
```kubectl get SecurityContextConstraints                                                                                                         11:08:11
NAME                              PRIV    CAPS         SELINUX     RUNASUSER          FSGROUP     SUPGROUP    PRIORITY     READONLYROOTFS   VOLUMES
anyuid                            false   <no value>   MustRunAs   RunAsAny           RunAsAny    RunAsAny    10           false            ["configMap","downwardAPI","emptyDir","persistentVolumeClaim","projected","secret"]
hostaccess                        false   <no value>   MustRunAs   MustRunAsRange     MustRunAs   RunAsAny    <no value>   false            ["configMap","downwardAPI","emptyDir","hostPath","persistentVolumeClaim","projected","secret"]
hostmount-anyuid                  false   <no value>   MustRunAs   RunAsAny           RunAsAny    RunAsAny    <no value>   false            ["configMap","downwardAPI","emptyDir","hostPath","nfs","persistentVolumeClaim","projected","secret"]
hostnetwork                       false   <no value>   MustRunAs   MustRunAsRange     MustRunAs   MustRunAs   <no value>   false            ["configMap","downwardAPI","emptyDir","persistentVolumeClaim","projected","secret"]
machine-api-termination-handler   false   <no value>   MustRunAs   RunAsAny           MustRunAs   MustRunAs   <no value>   false            ["downwardAPI","hostPath"]
node-exporter                     true    <no value>   RunAsAny    RunAsAny           RunAsAny    RunAsAny    <no value>   false            ["*"]
nonroot                           false   <no value>   MustRunAs   MustRunAsNonRoot   RunAsAny    RunAsAny    <no value>   false            ["configMap","downwardAPI","emptyDir","persistentVolumeClaim","projected","secret"]
privileged                        true    ["*"]        RunAsAny    RunAsAny           RunAsAny    RunAsAny    <no value>   false            ["*"]
restricted                        false   <no value>   MustRunAs   MustRunAsRange     MustRunAs   RunAsAny    <no value>   false            ["configMap","downwardAPI","emptyDir","persistentVolumeClaim","projected","secret"]
```

It looks like the `fsGroup` alone will do the job and gets it working as expected. Maybe we can revisit this later to create a dedicated `scc` and then tie our service account to it.

**Comment by veldpworld** [comment]
``` fatal: [localhost]: FAILED! => {"changed": false, "cmd": "/usr/bin/git clone --origin origin git@ssh.dev.azure.com:v3/dpwhotfsonline/IT%20Infra/AnsiblePlaybooks /var/lib/awx/projects/_10__azure_repo", "msg": "Cloning into '/var/lib/awx/projects/_10__azure_repo'...\nerror: chmod on /var/lib/awx/projects/_10__azure_repo/.git/config.lock failed: Operation not permitted\nfatal: could not set 'core.filemode' to 'false'", "rc": 128, "stderr": "Cloning into '/var/lib/awx/projects/_10__azure_repo'...\nerror: chmod on /var/lib/awx/projects/_10__azure_repo/.git/config.lock failed: Operation not permitted\nfatal: could not set 'core.filemode' to 'false'\n", "stderr_lines": ["Cloning into '/var/lib/awx/projects/_10__azure_repo'...", "error: chmod on /var/lib/awx/projects/_10__azure_repo/.git/config.lock ``` 

i followed the nfs volume, now it can write the files, but its giving the above error while sync the repo.
